### PR TITLE
Add cudf::strings:like function with multiple patterns

### LINKED ISSUE #10797: [FEA] `LIKE` support for a column of patterns, but a scalar escape character
**Is your feature request related to a problem? Please describe.**
I have to start off by apologizing. I know this is crazy difficult to do efficiently, but we have a customer that is asking for this. In SQL there is a string pattern matching system around `LIKE`.

Spark only supports `%` and `_` as special characters with an optional escape character passed in that defaults to `\`. 

In Spark if the pattern is a literal value, then it will parse it and translate the pattern into thinks like starts_with for `SOMETHING%`, ends_with for `%SOMETHING` or contains for `%SOMETHING%`. If it is anything else it is sent to the like operator.  For us we translate the pattern into a regular expression. `%` becomes `(.|\n)*` and `_` becomes `(.|\n)`. All other characters we escape before passing it to CUDF. I realize that this is super hard and it may need to be very slow to keep memory management under control. We get that.

**Describe the solution you'd like**
I would like to see a LIKE operator added to CUDF. Ideally it would take either a scalar pattern or a column of patterns along with an escape character and return a boolean column to say of the string matches the pattern.  If the pattern is null or the string is null the output should be null.

**Describe alternatives you've considered**
1. Write something ourselves
2. Go back to the customer and ask them what are they actually trying to do, and see if there is a better way to do it that does not involve this complex operator.

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Adds a `cudf::strings::like` function that accepts a column of patterns where each pattern is matched against the corresponding input string row. Only a single escape character is supported for all patterns.

Closes #10797 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/include/cudf/strings/contains.hpp [hunk 1]
```diff
@@ -238,6 +238,45 @@ std::unique_ptr<column> like(
   string_scalar const& escape_character = string_scalar(""),
   rmm::mr::device_memory_resource* mr   = rmm::mr::get_current_device_resource());
 
+/**
+ * @brief Returns a boolean column identifying rows which
+ * match the corresponding like pattern in the given patterns
+ *
+ * The like pattern expects only 2 wildcard special characters:
+ * - `%` any number of any character (including no characters)
```
Comment (robertmaynard) [inline cpp/include/cudf/strings/contains.hpp:246]:
```suggestion
 * - `%` zero or more of any character
```

File: cpp/src/strings/like.cu [hunk 2]
```diff
@@ -149,5 +182,14 @@ std::unique_ptr<column> like(strings_column_view const& input,
   return detail::like(input, pattern, escape_character, cudf::get_default_stream(), mr);
 }
 
+std::unique_ptr<column> like(strings_column_view const& input,
```
Comment (robertmaynard) [inline cpp/src/strings/like.cu:185]: We should add a test for when input and pattern are both zero length.

**Review by robertmaynard — CHANGES_REQUESTED** [review]

**Review by bdice — APPROVED** [review]
Very clean!

**Comment by davidwendt** [comment]
@gpucibot merge
